### PR TITLE
Add git clone support for WSGI

### PR DESCRIPTION
**Pull request by hxtmdev**
Adds a first read-only implementation of #232 when using WSGI (the dulwich integration didn't appear simple without it).

~Example use: `git clone https://example.com/_git_/user/calendars/e317e59d-9029-40e6-8335-13523c9d8b74`~  
Example use: `git clone https://example.com/user/calendars/e317e59d-9029-40e6-8335-13523c9d8b74/.git`

Resource is found by existing code, the repo is extracted and passed to dulwich.

Please let me know if this is the direction you had in mind.

[Tests](https://gitlab.com/hxtmdev/dav/-/blob/fa5a30c29909f860015956f390da18c8532c41a1/flake.nix#L170-L184) are currently only included in [the wrapper I just pushed](https://gitlab.com/hxtmdev/dav).
If the PR is otherwise okay please let me know in what form you'd like the test implementation in this PR.

### CODE REVIEW

**Comment by jelmer** [comment]
Thanks for working on this, this would be great to land. 

I think my only two comments here is that the path should probably be the other way around, i.e.:

```
/user/calendars/e317e59d-9029-40e6-8335-13523c9d8b74/.git
```

rather than:

```
/_git_/user/calendars/e317e59d-9029-40e6-8335-13523c9d8b74
```

so that we reuse the existing hierarchy

**Comment by hxtmdev** [comment]
Thanks for the feedback, changed the path to `/.git` suffix.

Two questions at this point
1. Does this need route prefix handling if it's WSGI-only?
2. I feel like this should have tests in this repository to prevent regressions. How do you want them? Something from the Python code in `/tests`? Some bash script from a GitHub action?

**Comment by jelmer** [comment]
Regarding tests, something basic in python would be great. Shell tests are less preferred since they're harder to run locally.

**Comment by jelmer** [comment]
I think ideally we'd actually interpret the path properly rather than just matching on ".git" anywhere in the path. I don't think that should have to block landing this PR as-is though.

File: xandikos/web.py [hunk 1]
```diff
@@ -1148,6 +1152,28 @@ def get_current_user_principal(env):
             ]
         )
 
+    async def _handle_request(self, request, environ, start_response=None):
+        # don't look any deeper than /user/calendars/foobar/.git
+        path_search_prefix = request.path.split(posixpath.sep)[:5];
```
Comment (jelmer) [inline xandikos/web.py:1157]: Some of the style checks appear to be failing.

**Review by jelmer — CHANGES_REQUESTED** [review]

**Comment by hxtmdev** [comment]
> Regarding tests, something basic in python would be great. Shell tests are less preferred since they're harder to run locally.

Alright, in that case I need some guidance I'm afraid.

From what I can tell
- `test_web.py` only goes up to the backend, the new code sits in the App though
- `test_webdav.py` does fake a WSGI deployment, should I put a minimal `StoreBasedCollection` into the test? Even then, should I start a whole TCP server to clone into a `MemoryRepo` or just verify that `/info/refs` changes?

Could you give me some pointers?

**Comment by jelmer** [comment]
Sorry for the slow reply, I had to page some of this back in. 

I think we need to do some more refactoring of this code for proper testing. 

For now, maybe you can just add a test to verify that the routing is done correctly, i.e. that the App will correctly return the contents of the "refs" file for a given repository?

**Comment by jelmer** [comment]
On a related note, I wonder if we could possibly drop the .git directory as well - just handle the appropriate paths under the calendar path directly. Maybe something for a follow-up PR as there are some potential traps there - what if you have a file called refs?

File: xandikos/web.py (hunk 1 above)
Comment (hxtmdev) [reply xandikos/web.py]: fixed `flake8`, `isort` seems to be red anyway?!

**Review by jelmer — APPROVED** [review]
Thanks!

**Pull request merged**